pkg/arvo/lib/contact-json.hoon

```diff
@@ -15,46 +15,62 @@
   |=  [pax=^path =contacts]
   ^-  [cord json]
   :-  (spat pax)
-  (contacts-to-json contacts)
+  (contacts-to-json pax contacts)
 ::
 ++  contacts-to-json
-  |=  con=contacts
+  |=  [=path con=contacts]
   ^-  json
-  =,  enjs:format
-  %-  pairs
+  %-  pairs:enjs:format
   %+  turn  ~(tap by con)
-  |=  [shp=^ship =contact]
+  |=  [=ship =contact]
   ^-  [cord json]
-  :-  (crip (slag 1 (scow %p shp)))
-  (contact-to-json contact)
+  [(crip (slag 1 (scow %p ship))) (contact-to-json path ship contact)]
 ::
 ++  contact-to-json
-  |=  con=contact
+  |=  [=path =ship con=contact]
   ^-  json
-  =,  enjs:format
-  %-  pairs
+  %-  pairs:enjs:format
   :~  [%nickname s+nickname.con]
       [%email s+email.con]
       [%phone s+phone.con]
       [%website s+website.con]
       [%notes s+notes.con]
       [%color s+(scot %ux color.con)]
-      [%avatar s+'TODO']
+      [%avatar (avatar-to-json path ship avatar.con)]
   ==
 ::
 ++  edit-to-json
-  |=  edit=edit-field
+  |=  [=path =ship edit=edit-field]
   ^-  json
-  =,  enjs:format
-  %+  frond  -.edit
+  %+  frond:enjs:format  -.edit
   ?-  -.edit
     %nickname  s+nickname.edit
     %email     s+email.edit
     %phone     s+phone.edit
     %website   s+website.edit
     %notes     s+notes.edit
     %color     s+(scot %ux color.edit)
-    %avatar    s+'TODO'
+    %avatar    (avatar-to-json path ship avatar.edit)
+  ==
+::
+++  avatar-to-json
+  |=  [=path =ship avat=(unit avatar)]
+  ^-  json
+  ?~  avat  ~
+  ?-  -.u.avat
+      %octt
+    :-  %s
+    %-  crip
+    %-  zing
+    :~  "/~groups/img"
+        (trip (spat path))
+        "/"
+        (trip (scot %p ship))
+        "."
+        (trip file-extension.u.avat)
+    ==
```

Comment:
Can you explain how the `%octt` case of `+avatar` is intended to be used? I'm confused.

Reply:
Yes. In the first version of this that we release, I’d like to only support %url on the front-end due to space-constraints in vere. However, ideally, you’d encode an image as base64, upload it to your urbit as a poke, it would be decoded from base64 in the %json mark, then stored in contact-store. When converting back to %json, you pass a relative URL through which the binary file can be downloaded by the client. I’ve tested this whole flow and it works, just want to hide it away for a little while until the space-constraints are less salient.

Reply:
I see. Thanks!

Might want to consider just leaving out this case entirely from implementation, for the time being. "Half-in and intentionally not in use" isn't the best state to put things in imo.

Also, when this does go in, we may want to write these to clay instead of storing in state. Both to provide savvy users more control, and to play nicely with standardized filesystem serving that Eyre's slated to get.

(We'll probably want to impose a limit on the size of avatars we accept regardless of memory pressure?)

Reply:
Disagree, as I view Clay’s future as being a filesystem exclusively for source files.

Reply:
I also think it’s a good experimental capability for us to have, and disabling from the front-end is just fine.